docs/build/voting-dapp-tutorial.md

```diff
@@ -0,0 +1,428 @@
+
+## Smart Contracts
+
+Our voting app should allow you to submit a vote while not allowing an address to vote more than once.
+Additionally we will use a *gas station* to fund the gas fees for interacting with our contract, meaning our users don't have to worry about paying gas.
+
+### Voting
+
+Let's start to implement our voting contract.
+
+Assuming you are still in your project directory, create the `vote.pact` and `vote.repl` files:
+
+```clojure
+touch pact/vote.pact pact/vote.repl
+```
+
+`vote.pact` is the contract source code while in `vote.repl` we'll write tests.
+
+Copy the following code in the `vote.pact` file:
+
+```clojure
+(define-keyset 'vote-admin-keyset)
+
+(module simple-vote 'vote-admin-keyset
+  "Simple voting module"
+
+  (defun vote (key)
+    "Submit a new vote"
+    (format "Voted {}!" [key]))
+)
+
+(vote "A")
+```
+
+This snippet defines a module "simple-vote" that holds a function "vote" that takes a parameter and displays a formatted string that includes the parameter.
+
+Before we move forward, let's quickly test our code. Copy the code below in the `vote.repl` file:
+
+```clojure
+;; begin a transaction
+(begin-tx)
+;; set transaction signature key to my-key
+(env-keys ["my-key"])
+;; set environment data to the admin-keyset with keys my-key and predicate function of keys-all
+(env-data { 'vote-admin-keyset: { "keys": ["my-key"], "pred": "keys-all" } })
+;; load vote.pact into the REPL
+(load "vote.pact")
+;; commit the transaction
+(commit-tx)
+```
+
+To run the test type the following command in your terminal:
+```clojure
+$ pact
+pact> (load "vote.repl")
+```
+And you should see an output similar to the one below (the hash will be different):
+```clojure
+"Loading vote.repl..."
+"Begin Tx 0"
+"Setting transaction keys"
+"Setting transaction data"
+"Loading vote.pact..."
+"Keyset defined"
+"Loaded module simple-vote, hash jOBJiSEH5HVsAmLNSdHq_D3Kl6TkFpCJcEWPtJD86Gc"
+"Voted A!"
+"Commit Tx 0"
+```
+
+Let's continue by adding a bit more logic. Copy this snippet just below the module declaration:
+
+```clojure
+  (defschema votes-schema
+    "Votes table schema"
+    option:string
+    count:integer)
+
+  (deftable votes:{votes-schema})
+  ```
+In the snippet we've got 2 new declaration:
+* `defschema votes-schema` -> defines the schema for our table where we will keep track of the number of votes
+* `deftable votes:{votes-schema}` -> defines the `votes` table that will use the schema defined above
+
+Now we need to update our `vote` function:
+
+```clojure
+(defun vote (key:string)
+    "Submit a new vote"
+    (enforce (contains key ["A" "B"]) "You can only vote for A or B")
+    (with-read votes key { "count" := count }
+      (update votes key { "count": (+ count 1) })
+    )
+    (format "Voted {}!" [key])
+  )
+```
+
+Again, quick recap of what's happening above:
+* we use `enforce` to ensure that only the given list of vote keys are allowed
```

Comment:
can we rephrase this ? Or better yet, come up with a better way than submitting raw strings as votes

Reply:
This was refactored.